### PR TITLE
[`cuda kernels`] only compile them when initializing

### LINKED ISSUE #29130: Move kernel compilation to init rather than at import stage
### Feature request

Some models like Deformable DETR rely on custom CUDA kernels to be compiled as seen [here](https://github.com/huggingface/transformers/blob/f7ef7cec6c6c162087421f36a17eabdbb223579d/src/transformers/models/deformable_detr/modeling_deformable_detr.py#L54).

Currently these are compiled when importing the Transformers library, but this needs to happen later, when initializing the models.

All custom CUDA kernels are defined here: https://github.com/huggingface/transformers/tree/main/src/transformers/kernels

### Motivation

This is pretty important to fix actually as currently running `make fixup` on machines that have `ninja` installed will compile all these kernels before running the quality checks, making it super slow.. thanks @younesbelkada for the info

### Your contribution

Not sure I can help with this, the current workaround is simply removing `ninja` from the environment

### PR DESCRIPTION
**Pull request by ArthurZucker**
# What does this PR do?
Fixes #29130, from 1min to 6seconds

### CODE REVIEW

File: src/transformers/models/deformable_detr/modeling_deformable_detr.py [hunk 1]
```diff
@@ -46,21 +48,42 @@
...
+        with_cuda=True,
+        extra_include_paths=[str(root)],
+        extra_cflags=["-DWITH_CUDA=1"],
+        extra_cuda_cflags=[
+            "-DCUDA_HAS_FP16=1",
+            "-D__CUDA_NO_HALF_OPERATORS__",
+            "-D__CUDA_NO_HALF_CONVERSIONS__",
+            "-D__CUDA_NO_HALF2_OPERATORS__",
+        ],
+    )
```
Comment (ArthurZucker) [inline src/transformers/models/deformable_detr/modeling_deformable_detr.py:85]:
```suggestion
    load(
        MultiScaleDeformableAttention = "MultiScaleDeformableAttention",
        src_files,
        with_cuda=True,
        extra_include_paths=[str(root)],
        extra_cflags=["-DWITH_CUDA=1"],
        extra_cuda_cflags=[
            "-DCUDA_HAS_FP16=1",
            "-D__CUDA_NO_HALF_OPERATORS__",
            "-D__CUDA_NO_HALF_CONVERSIONS__",
            "-D__CUDA_NO_HALF2_OPERATORS__",
        ],
    )
```

File: src/transformers/models/deta/modeling_deta.py [hunk 2]
```diff
@@ -50,10 +50,18 @@
...
 
+
+# Copied from models.deformable_detr.load_cuda_kernels
 def load_cuda_kernels():
     from torch.utils.cpp_extension import load
 
+    global MultiScaleDeformableAttention
+    # Only load the kernel if it's not been loaded yet or if we changed the context length
+    if MultiScaleDeformableAttention is not None:
+        return
```
Comment (ArthurZucker) [inline src/transformers/models/deta/modeling_deta.py:63]:
```suggestion
```

**Review by amyeroberts — APPROVED** [review]
Great - thanks for fixing!

As long as the model tests pass in an env with ninja installed I think it's good to go!

**Comment by ArthurZucker** [comment]
I'll make sure of that before merging! Testing now!

**Comment by ArthurZucker** [comment]
```bash
FAILED tests/models/deformable_detr/test_modeling_deformable_detr.py::DeformableDetrModelIntegrationTests::test_inference_object_detection_head - AssertionError: False is not true
FAILED tests/models/deformable_detr/test_modeling_deformable_detr.py::DeformableDetrModelIntegrationTests::test_inference_object_detection_head_equivalence_cpu_gpu - AssertionError: assert False
FAILED tests/models/deformable_detr/test_modeling_deformable_detr.py::DeformableDetrModelIntegrationTests::test_inference_object_detection_head_with_box_refine_two_stage - AssertionError: False is not true
``` failing. Logits do no match (probably same on main)

```python 
FAILED tests/models/deta/test_modeling_deta.py::DetaModelIntegrationTests::test_inference_object_detection_head - AssertionError: False is not true
FAILED tests/models/deta/test_modeling_deta.py::DetaModelIntegrationTests::test_inference_object_detection_head_swin_backbone - AssertionError: False is not true
``` failing as well 

Yoso is alright

File: src/transformers/models/deformable_detr/modeling_deformable_detr.py [hunk 3]
```diff
@@ -46,21 +48,42 @@
...
-    MultiScaleDeformableAttention = None
+MultiScaleDeformableAttention = None
+
+
+def load_cuda_kernels():
+    from torch.utils.cpp_extension import load
+
+    global MultiScaleDeformableAttention
+
+    root = Path(__file__).resolve().parent.parent.parent / "kernels" / "deta"
```
Comment (EduardoPach) [inline src/transformers/models/deformable_detr/modeling_deformable_detr.py:63]: Hey @ArthurZucker, was it intended for the root to be `deta` related?

Comment (amyeroberts) [reply src/transformers/models/deformable_detr/modeling_deformable_detr.py:63]: @EduardoPach I think this is a copy-pasta. Would you like to open a PR to fix?

Comment (EduardoPach) [reply src/transformers/models/deformable_detr/modeling_deformable_detr.py:63]: @amyeroberts, sure!

Comment (ArthurZucker) [reply src/transformers/models/deformable_detr/modeling_deformable_detr.py:63]: thanks

**Pull request merged**